CMakeLists.txt

```
@@ -48,6 +48,14 @@ else ()
     message(STATUS "ccache will not be used")
 endif ()
 
+# Prefer forced colored compiler output if there's a chance we'd otherwise get none at all.
+# Ninja and ccache "consume" the compiler output, breaking the terminal detection of compilers.
+# ccache tries to solve the problem, but can only do so if it determines that it's calling GCC or Clang. It uses a very lightweight heuristic, which breaks easily.
+if ((CMAKE_GENERATOR STREQUAL "Ninja" OR (CCACHE_PROGRAM AND USE_CCACHE)) AND NOT DEFINED CMAKE_COLOR_DIAGNOSTICS)
```

```suggestion
if ( NOT DEFINED CMAKE_COLOR_DIAGNOSTICS)
```
Just always default to colored output but allow to turn it off in the standard cmake way?

---

If tried this before and it'd be a very unfortunate outcome. GCC and Clang are designed to do the right thing automatically without getting told from outside. I.e. if I build from vim or pipe output from make into a file, the compiler outputs no color codes. But if I build in a terminal (also from a terminal inside vim) I get color output. Except when I force it off via `CMAKE_COLOR_DIAGNOSTICS=OFF`.

The "problem" when running as GitHub action is that the compiler doesn't see a terminal. So in order to see colored output, yes one has to force it to `ON`, that's the technically correct solution here.